### DISCUSSION TITLE
How to reuse existing edge in combination with new geometry?

### DISCUSSION THREAD

**Discussion by wfraser** [CadQuery/cadquery#1688] [Q&A] [answered]
Say I've got some solid with a complex shape, for example:
```
box = (
    cq.Workplane("XY")
    .box(10, 10, 1)
    .faces(">Z")
    .rect(3, 3, forConstruction=True)
    .vertices()
    .hole(1)
)
```
![image](https://github.com/user-attachments/assets/ce50edf0-4b2f-4b8b-bc1a-e739e8cbbb7b)

I want to use the outer edge combined with some other shape (say, a circle) and extrude that.

In this example the outer edge is just a square and I can easily just draw it again:
```
desiredShape = (
    box.faces(">Z")
    .rect(10, 10)
    .circle(4)
    .extrude(0.5, combine=False)
)
```
![image](https://github.com/user-attachments/assets/84039ad5-8af1-4c15-8374-2a6b394fabad)

But what if it's actually a complex shape and I really don't want to re-define it?

I know how to select the outer edge, by finding the longest one:
```
box.faces(">Z").edges().sort(lambda edge: edge.Length() * -1).first()
```
but when I try to use that in place of the `.rect(10, 10)` in the previous block, it's seemingly ignored, and just extrudes the inner circle instead:

```
myShape = (
    box.faces(">Z")
    .edges()
    .sort(lambda edge: edge.Length()*-1)
    .first()
    .circle(4)
    .extrude(0.5, combine=False)
)
```
![image](https://github.com/user-attachments/assets/21c508e4-ed27-4757-8df2-bf481f24b081)

If I remove the call to `.circle()`, I get `No pending wires present`. I tried adding `.toPending()` which had no effect.

How do I reuse this existing edge in combination with other new geometry?

**Comment by lorenzncode** [comment, marked as answer]
Here are a few ideas.  You were close with (1).
 
1.  Solution with Workplane, toPending.  Try selecting the wire not the edges:

```py
# this matches the desired shape
myShape2 = (
    box.faces(">Z")
    .wires()
    .sort(lambda wire: wire.Length() * -1)
    .first()
    .toPending()
    .circle(4)
    .extrude(0.5, combine=False)
)
```


2. With Sketch. You can reuse the sketch for the original box and later for myShape.

```py
import cadquery as cq

sk1 = cq.Sketch().rect(10, 10)
sk2 = sk1.copy().rarray(3, 3, 2, 2).circle(0.5, mode="s")

box1 = cq.Workplane().placeSketch(sk2).extrude(1)

# for comparison with myShape
desiredShape = box1.faces(">Z").rect(10, 10).circle(4).extrude(0.5, combine=False)

myShape = (
    box1.faces(">Z")
    .placeSketch(sk1.copy().circle(4, mode="s"))
    .extrude(0.5, combine=False)
)
```


3. With free func API. 

```py
import cadquery as cq
from cadquery.occ_impl.shapes import *

# name the box as box1 so it does not conflict with box() func
box1 = (
    cq.Workplane("XY")
    .box(10, 10, 1)
    .faces(">Z")
    .rect(3, 3, forConstruction=True)
    .vertices()
    .hole(1)
)

# for comparison with myShape
desiredShape = box1.faces(">Z").rect(10, 10).circle(4).extrude(0.5, combine=False)

wire1 = box1.faces(">Z").wires().sort(lambda wire: wire.Length() * -1).first().val()
wire2 = box1.faces(">Z").circle(4).val()  # new circle
face1 = face(wire1) - face(wire2)  # new face for desired shape
myShape = extrude(face1, (0, 0, 0.5))  # the desired shape
```

edit - remove issue link

**Reply by wfraser** [reply]
Ahh, thank you Lorenz! It seems I don't understand the distinction between edges and wires well enough yet. That was what I was missing.

**Reply by lorenzncode** [reply]
Regarding edges vs wires, see also the OCCT Introduction section for a description: 
https://dev.opencascade.org/doc/overview/html/index.html#intro_overview_moddata


_Edge – a shape corresponding to a curve and bounded by a vertex at each extremity;
Wire – a sequence of edges connected by their vertices;_


Example build wire from edges:

```py
from cadquery.occ_impl.shapes import segment, spline, wire

# create two edges
e1 = segment((0, 0), (10, 0))
e2 = spline([(10, 0), (0, 10)], [(1, 0), (0, -1)])

# create a wire
w1 = wire(e1, e2)

# create a wire with closing edge; three total edges
w2 = w1.close()
```
